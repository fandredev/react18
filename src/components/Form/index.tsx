import React, { Dispatch, FormEvent, useState } from 'react';
import { TasksListProps } from "types/tasks";
import Button from "../Button";
import styles from "./form.module.scss";

import { v4 as uuidv4 } from 'uuid';

interface FormProps {
  setTasks: Dispatch<React.SetStateAction<TasksListProps[]>>
}

export default function Form({ setTasks }: FormProps) {
  const [task, setTask] = useState("")
  const [time, setTime] = useState("00:00:00")

  function addNewTask(event: FormEvent<HTMLFormElement>) {
    event.preventDefault()

    setTasks((oldState) => [
      ...oldState,
      { task, time, selected: false, completed: false, id: uuidv4() }
    ])

    clearForm()
  }

  function clearForm() {
    setTask('')
    setTime('00:00:00')
  }

  return (
    <form className={styles.newTaskForm} onSubmit={event => addNewTask(event)}>
      <div className={styles.inputContainer}>
        <label htmlFor="task">
          Adicione um novo estudo
        </label>
        <input onChange={event => setTask(event.target.value)} value={task} type="text" name="task" id="task"
          placeholder="O que você quer estudar" required />
      </div>
      <div className={styles.inputContainer}>
        <label htmlFor="time">
          Tempo
        </label>
        <input data-testid="input-time" onChange={event => setTime(event.target.value)} value={time} type="time" step="1" name="time" id="time"
          min="00:00:00" max="01:30:00" required />
      </div>
      <Button type="submit">Adicionar</Button>
    </form>
  )
}

// export default class Form extends Component<FormProps> {
//   state = {
//     task: '',
//     time: '00:00:00'
//   }
//
//   protected changeValueTask(event: ChangeEvent<HTMLInputElement>) {
//     const task = event.target.value
//     this.setState({
//       ...this.state,
//       task
//     })
//   }
//
//   protected changeValueTime(event: ChangeEvent<HTMLInputElement>) {
//     const time = event.target.value
//     this.setState({
//       ...this.state,
//       time
//     })
//   }
//
//   protected addNewTask(event: FormEvent<HTMLFormElement>) {
//     event.preventDefault()
//     this.props.setTasks((oldState) => [
//       ...oldState,
//       {...this.state, selected: false, completed: false, id: uuidv4()}
//     ])
//
//     this.clearForm()
//   }
//
//   private clearForm() {
//     this.setState({
//       task: '',
//       time: '00:00:00'
//     })
//   }
//
//   render() {
//     const {time, task} = this.state
//     return (
//       <form className={styles.newTaskForm} onSubmit={event => this.addNewTask(event)}>
//         <div className={styles.inputContainer}>
//           <label htmlFor="task">
//             Adicione um novo estudo
//           </label>
//           <input onChange={event => this.changeValueTask(event)} value={task} type="text" name="task" id="task"
//                  placeholder="O que você quer estudar" required/>
//         </div>
//         <div className={styles.inputContainer}>
//           <label htmlFor="time">
//             Tempo
//           </label>
//           <input onChange={event => this.changeValueTime(event)} value={time} type="time" step="1" name="time" id="time"
//                  min="00:00:00" max="01:30:00" required/>
//         </div>
//         <Button type="submit">Adicionar</Button>
//       </form>
//     )
//   }
// }
