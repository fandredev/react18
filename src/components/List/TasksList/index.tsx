import classNames from 'classnames';
import { TasksListProps } from "types/tasks";
import styles from './taskslist.module.scss';

interface ListTasksProps extends TasksListProps {
  selectTask: (selectedTask: TasksListProps) => void
}


export default function TasksList({ task, time, id, selected, completed, selectTask }: ListTasksProps) {
  return (
    <li className={classNames({
      [styles.item]: true,
      [styles.listSelected]: selected,
      [styles.listCompleted]: completed
    })} onClick={() => !completed && selectTask({
      task, time, id, selected, completed
    })}>
      <h3>{task}</h3>
      <span>{time}</span>
      {completed && <span className={styles.completed} aria-label="Tarefa completada" />}
    </li>
  )
}
