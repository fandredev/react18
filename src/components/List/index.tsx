import { TasksListProps } from "types/tasks";
import styles from "./list.module.scss";
import TasksList from "./TasksList";

interface ListProps {
  tasks: TasksListProps[]
  selectTask: (selectedTask: TasksListProps) => void
}

export default function List({ tasks, selectTask }: ListProps): JSX.Element {

  return (
    <aside className={styles.listTasks}>
      {tasks.length === 0 ? <h2>Você ainda não adicionou algum estudo hoje.</h2> : <h2>Estudos do dia</h2>}
      <ul>
        {tasks.map((task) =>
          <TasksList selectTask={selectTask} {...task} key={task.id} />
        )}
      </ul>
    </aside>
  )
}
