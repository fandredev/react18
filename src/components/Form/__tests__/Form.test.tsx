import { fireEvent, render, screen } from '@testing-library/react';
import { task } from 'common/mocks/List.mock';
import Form from '..';

describe(`Test Component: ${Form.name}`, () => {

  beforeEach(() => {
    render(<Form setTasks={() => [{ ...task }]} />);
  })
  it('Should be possible to snapshot the Button component when the page is rendered', () => {
    const { container } = render(<Form setTasks={() => [{ ...task }]} />);
    expect(container).toMatchSnapshot();
  });

  it('Deve ser possível digitar um novo tipo de estudo no input de estudo do formulario', () => {

    const inputNewTask = screen.getByPlaceholderText('O que você quer estudar');
    fireEvent.change(inputNewTask, {
      target: {
        value: 'Estudar React',
      },
    });

    expect(inputNewTask.getAttribute('value')).toBe('Estudar React');
  });

  it('Deve ser possível digitar um novo tempo para o estudo no input de tempo do formulario', () => {
    const inputNewTaskTime = screen.getByTestId('input-time');
    expect(inputNewTaskTime).toHaveValue('00:00:00');

    fireEvent.change(inputNewTaskTime, {
      target: {
        value: '01:02:00',
      },
    });

    expect(inputNewTaskTime.getAttribute('value')).toBe('01:02:00');
  });
});
