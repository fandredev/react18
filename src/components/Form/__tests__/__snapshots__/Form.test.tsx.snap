// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Component: Form Should be possible to snapshot the Button component when the page is rendered 1`] = `
<div>
  <form
    class="newTaskForm"
  >
    <div
      class="inputContainer"
    >
      <label
        for="task"
      >
        Adicione um novo estudo
      </label>
      <input
        id="task"
        name="task"
        placeholder="O que vocÃª quer estudar"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="inputContainer"
    >
      <label
        for="time"
      >
        Tempo
      </label>
      <input
        data-testid="input-time"
        id="time"
        max="01:30:00"
        min="00:00:00"
        name="time"
        required=""
        step="1"
        type="time"
        value="00:00:00"
      />
    </div>
    <button
      class="button"
      type="submit"
    >
      Adicionar
    </button>
  </form>
</div>
`;
